### ISSUE TITLE
Dots stop and "twitch"

### ISSUE THREAD

**Issue by pserwylo** [babydots/babydots#49]
Feedback received via email from a Google Play user (via Google Translate and paraphrased):

> There is one problem - after about a minute, one and a half points stop in place and begin to twitch.
> Occurs on a OnePlus 8t smartphone with battery saver and smart mode disabled. Works fine on a xiaomi mi 10t pro.

Unfortunately I cannot reproduce this, but I will read through the code to try and see what might be causing it.

If anybody else is able to reproduce, especially if you are able to obtain a logcat, could you please let us know below? That would greatly help the effort to fix this issue.

**Comment by lateski** [comment]
I have OnePlus Nord with Android 11 and I have this exactly same issue. The issue seems to be caused by a memory leak in graphics side. About 2 minutes after opening the app it crashes with following logcat.

```
15326-455/com.serwylo.babydots D/skia: Could not allocate vertices
15326-455/com.serwylo.babydots W/Adreno-GSL: <sharedmem_gpuobj_alloc:2736>: sharedmem_gpumem_alloc: mmap failed errno 12 Out of memory
15326-455/com.serwylo.babydots E/Adreno-GSL: <gsl_memory_alloc_pure:2604>: GSL MEM ERROR: kgsl_sharedmem_alloc ioctl failed.
...
15326-455/com.serwylo.babydots W/Adreno-GSL: <sharedmem_gpuobj_alloc:2736>: sharedmem_gpumem_alloc: mmap failed errno 12 Out of memory
15326-455/com.serwylo.babydots E/Adreno-GSL: <gsl_memory_alloc_pure:2604>: GSL MEM ERROR: kgsl_sharedmem_alloc ioctl failed.
15326-455/com.serwylo.babydots A/libc: Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x79131c6000 in tid 455 (RenderThread), pid 15326 (erwylo.babydots)

```
As seen from this image the application hogs quite much memory on the graphics side. After it reaches about 550MB it gets OOM
![Selection_128](https://user-images.githubusercontent.com/1245058/140656300-2c05e0c9-e720-43bd-9aaa-431f1b2f6ed0.png)

I tried googling this issue and found this: [Android canvas.draw leaks graphic memory](https://stackoverflow.com/a/67702834) which suggests that using **Paint.Style.STROKE** might be causing this memory leak. I tried commenting out the drawing of dot borders in the AnimatedDots.onDraw method because it was the only draw call using stroke paint style. After that I ran babydots for 10 minutes(!) and it didn't crash or leak memory anymore.

**Comment by pserwylo** [comment]
Ah, thank you so very much for this wonderful analysis. I think this is one of the more detailed comments on a bug report in one of my repos!

Sounds like the solution may be to ditch the `Paint.Style.STROKE` and instead we can simulate this with two different calls to `Paint.Style.FILL` - the outer one the colour of the stroke and slightly larger, and the inner one painted on top to simulate a stroke.

Pull requests are of course welcome if you'd like to have a go at fixing - but if not, I'll try to get to it in the next week or two.

**Comment by lateski** [comment]
So I wasn't able to find any better solutions so I implemented what you proposed. I ran the software for 20 minutes with Android profiler on and I wasn't able to see leaks anymore. I don't have any other devices currently so I couldn't compare with anything.
I commented out and left the old stroke paint definitions into init as this new solution is a bit unusual and should probably be replaced with something else. Proper solution would probably require more debugging with the `drawCircle()` and some fiddling with different flags `Canvas` offers. But I think this solutions is somewhat universal.
